File: ruby/examples/basic_example.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Copyright (c) Microsoft Corporation. All rights reserved.

# Basic example of using the Copilot Ruby SDK.
#
# Prerequisites:
#   1. Install the Copilot CLI (https://github.com/github/copilot)
#   2. Authenticate via `copilot auth login` or set GITHUB_TOKEN
#
# Usage:
#   ruby examples/basic_example.rb

require_relative "../lib/copilot"

# ------------------------------------------------------------------
# 1. Define custom tools
# ------------------------------------------------------------------

weather_tool = Copilot.define_tool(
  name: "get_weather",
  description: "Get the current weather for a city",
  parameters: {
    type: "object",
    properties: {
      city: { type: "string", description: "The city name" },
      unit: { type: "string", enum: %w[celsius fahrenheit], description: "Temperature unit" },
    },
    required: ["city"],
  }
) do |args, _invocation|
  city = args["city"] || args[:city]
  unit = args["unit"] || args[:unit] || "celsius"
  # In a real tool you'd call a weather API here.
  "The weather in #{city} is 22 degrees #{unit} and sunny."
end

calculator_tool = Copilot.define_tool(
  name: "calculator",
  description: "Evaluate a simple mathematical expression",
  parameters: {
    type: "object",
    properties: {
      expression: { type: "string", description: "Math expression to evaluate" },
    },
    required: ["expression"],
  }
) do |args, _invocation|
  expr = args["expression"] || args[:expression]
  # WARNING: eval is used here for demonstration only. Never eval untrusted input in production.
  result = eval(expr) # rubocop:disable Security/Eval
  Copilot::ToolResult.new(
    text_result_for_llm: "Result: #{result}",
    result_type: Copilot::ToolResultType::SUCCESS,
    tool_telemetry: { expression: expr, result: result }
  )
end

# ------------------------------------------------------------------
# 2. Create client and session
# ------------------------------------------------------------------

# Set the path to your Copilot CLI binary. Adjust as needed.
cli_path = ENV["COPILOT_CLI_PATH"] || "copilot"

puts "Creating Copilot client..."
client = Copilot::CopilotClient.new(
  cli_path: cli_path,
  log_level: "info"
)

begin
  client.start
  puts "Connected! State: #{client.state}"

  # Ping the server
  ping_response = client.ping("hello from Ruby SDK")
  puts "Ping response: #{ping_response.message} (protocol v#{ping_response.protocol_version})"

  # List available models
  models = client.list_models
  puts "Available models: #{models.map(&:id).join(', ')}"

  # Create a session with tools
  puts "\nCreating session..."
  session = client.create_session(
    tools: [weather_tool, calculator_tool],
    on_permission_request: ->(request, ctx) {
      puts "Permission requested: #{request.kind}"
      # Auto-approve for this example
      Copilot::PermissionRequestResult.new(kind: Copilot::PermissionKind::APPROVED)
    },
    on_user_input_request: ->(request, ctx) {
      puts "Agent asks: #{request.question}"
      print "> "
      answer = $stdin.gets&.chomp || ""
      Copilot::UserInputResponse.new(answer: answer, was_freeform: true)
    }
  )
  puts "Session created: #{session.session_id}"

  # Subscribe to events
  unsub = session.on do |event|
    case event.type
    when Copilot::SessionEventType::ASSISTANT_MESSAGE
      content = event.data["content"] || event.data[:content]
      puts "\nAssistant: #{content}"
    when Copilot::SessionEventType::TOOL_EXECUTION_START
      tool = event.data["toolName"] || event.data[:toolName]
      puts "  [Tool: #{tool}]"
    when Copilot::SessionEventType::TOOL_EXECUTION_COMPLETE
      success = event.data["success"]
      puts "  [Tool complete: #{success ? 'OK' : 'FAILED'}]"
    when Copilot::SessionEventType::SESSION_ERROR
      msg = event.data["message"] || event.data[:message]
      puts "  [ERROR: #{msg}]"
    end
  end

  # ------------------------------------------------------------------
  # 3. Send messages
  # ------------------------------------------------------------------

  puts "\nSending: What's the weather in Tokyo?"
  response = session.send_and_wait(prompt: "What's the weather in Tokyo?", timeout: 120)
  if response
    puts "Final response received."
  end

  puts "\nSending: What is 42 * 17 + 3?"
  response = session.send_and_wait(prompt: "What is 42 * 17 + 3?", timeout: 120)
  if response
    puts "Final response received."
  end

  # ------------------------------------------------------------------
  # 4. Retrieve history
  # ------------------------------------------------------------------

  puts "\n--- Session History ---"
  messages = session.get_messages
  messages.each do |evt|
    puts "  [#{evt.type}] #{evt.id}"
  end

  # ------------------------------------------------------------------
  # 5. v2.0 Features
  # ------------------------------------------------------------------

  # Session Metadata
  meta = client.get_session_metadata(session.session_id)
  puts "Session ID: #{meta.session_id}" if meta

  # Skills (uncomment to use)
  # skill_session = client.create_session(
  #   skill_directories: ["./skills"],
  #   include_sub_agent_streaming_events: true
  # )

  # ------------------------------------------------------------------
  # 6. Cleanup
  # ------------------------------------------------------------------

  unsub.call
  session.destroy
  puts "\nSession destroyed."

rescue StandardError => e
  puts "Error: #{e.class}: #{e.message}"
  puts e.backtrace&.first(5)&.join("\n")
ensure
  errors = client.stop
  unless errors.empty?
    puts "Cleanup errors:"
    errors.each { |err| puts "  - #{err.message}" }
  end
  puts "Client stopped."
end

File: perl/examples/basic_example.pl
#!/usr/bin/env perl
# Copyright (c) Microsoft Corporation. All rights reserved.
#
# basic_example.pl - Demonstrates basic usage of the GitHub Copilot Perl SDK
#
# Prerequisites:
#   1. Install the Copilot CLI (e.g., npm install -g @github/copilot)
#   2. Authenticate: copilot auth login
#   3. Install Perl dependencies:
#      cpanm Moo JSON::PP UUID::Tiny File::Which Thread::Queue
#
# Usage:
#   perl basic_example.pl
#   perl basic_example.pl --cli-path /path/to/copilot

use strict;
use warnings;
use FindBin;
use lib "$FindBin::Bin/../lib";

use GitHub::Copilot::Client;
use GitHub::Copilot::Types;
use GitHub::Copilot::DefineTool qw(define_tool);

# ---------------------------------------------------------------------------
# Parse command-line options
# ---------------------------------------------------------------------------

my $cli_path;
for my $i (0 .. $#ARGV) {
    if ($ARGV[$i] eq '--cli-path' && defined $ARGV[$i + 1]) {
        $cli_path = $ARGV[$i + 1];
    }
}

# ---------------------------------------------------------------------------
# Define a custom tool
# ---------------------------------------------------------------------------

my $weather_tool = define_tool(
    name        => 'get_weather',
    description => 'Get the current weather for a location',
    parameters  => {
        type       => 'object',
        properties => {
            location => {
                type        => 'string',
                description => 'City name or location',
            },
        },
        required => ['location'],
    },
    handler => sub {
        my ($args, $invocation) = @_;
        my $location = $args->{location} // 'unknown';
        # Simulated weather response
        return "Currently 72 degrees Fahrenheit and sunny in $location.";
    },
);

# ---------------------------------------------------------------------------
# Create client and session
# ---------------------------------------------------------------------------

my %client_args;
$client_args{cli_path} = $cli_path if defined $cli_path;

my $client = GitHub::Copilot::Client->new(%client_args);

print "Starting Copilot client...\n";
$client->start();
print "Connected! State: " . $client->get_state() . "\n";

# Ping the server
my $ping = $client->ping("hello from Perl");
print "Ping response: " . $ping->message . "\n";
print "Protocol version: " . ($ping->protocolVersion // 'N/A') . "\n";

# List available models
print "\nAvailable models:\n";
my $models = $client->list_models();
for my $model (@$models) {
    print "  - " . $model->id . " (" . $model->name . ")\n";
}

# Create a session with the custom tool
print "\nCreating session...\n";
my $session = $client->create_session(
    GitHub::Copilot::Types::SessionConfig->new(
        tools => [$weather_tool],
        on_permission_request => sub {
            my ($request, $context) = @_;
            print "[Permission] Approving: $request->{kind}\n";
            return { kind => 'approved' };
        },
    ),
);
print "Session created: " . $session->session_id . "\n";

# Subscribe to events
my $unsubscribe = $session->on(sub {
    my ($event) = @_;
    my $type = $event->type;

    if ($type eq 'assistant.message') {
        my $content = $event->data->{content} // '';
        print "\n[Assistant] $content\n";
    } elsif ($type eq 'assistant.message_delta') {
        # Streaming delta
        my $delta = $event->data->{deltaContent} // '';
        print $delta;
    } elsif ($type eq 'tool.executing') {
        my $tool_name = $event->data->{toolName} // 'unknown';
        print "[Tool] Executing: $tool_name\n";
    } elsif ($type eq 'session.idle') {
        print "[Session] Idle\n";
    } elsif ($type eq 'session.error') {
        my $msg = $event->data->{message} // 'unknown error';
        print "[Error] $msg\n";
    }
});

# ---------------------------------------------------------------------------
# Send a message and wait for the response
# ---------------------------------------------------------------------------

print "\nSending message: 'What is the weather in Seattle?'\n";
print "-" x 60 . "\n";

my $response = $session->send_and_wait(
    GitHub::Copilot::Types::MessageOptions->new(
        prompt => 'What is the weather in Seattle?',
    ),
    120,  # 120 second timeout
);

if ($response) {
    print "\n" . "=" x 60 . "\n";
    print "Final response type: " . $response->type . "\n";
    my $content = $response->data->{content} // '(no content)';
    print "Final content: $content\n";
} else {
    print "No assistant response received.\n";
}

# ---------------------------------------------------------------------------
# Send another message (no tool needed)
# ---------------------------------------------------------------------------

print "\n" . "-" x 60 . "\n";
print "Sending message: 'What is 2+2?'\n";
print "-" x 60 . "\n";

my $response2 = $session->send_and_wait(
    { prompt => 'What is 2+2? Reply with just the number.' },
    60,
);

if ($response2) {
    my $content = $response2->data->{content} // '(no content)';
    print "Answer: $content\n";
}

# ---------------------------------------------------------------------------
# Retrieve conversation history
# ---------------------------------------------------------------------------

print "\nConversation history:\n";
my $messages = $session->get_messages();
for my $event (@$messages) {
    my $type = $event->type;
    if ($type eq 'user.message') {
        print "  [User] " . ($event->data->{content} // '') . "\n";
    } elsif ($type eq 'assistant.message') {
        my $content = $event->data->{content} // '';
        my $preview = length($content) > 80 ? substr($content, 0, 80) . '...' : $content;
        print "  [Assistant] $preview\n";
    }
}

# --- v2.0 Features ---

# Session Metadata
my $meta = $client->get_session_metadata($session->session_id);
if ($meta) {
    print "Session ID: " . $meta->{session_id} . "\n";
}

# Skills (uncomment to use)
# my $skill_session = $client->create_session(
#     skill_directories => ['./skills'],
#     include_sub_agent_streaming_events => 1,
# );

# ---------------------------------------------------------------------------
# Cleanup
# ---------------------------------------------------------------------------

print "\nCleaning up...\n";
$unsubscribe->();
$session->destroy();
my $errors = $client->stop();
if (@$errors) {
    print "Cleanup errors:\n";
    for my $e (@$errors) {
        print "  - $e\n";
    }
}
print "Done.\n";
